cql3/restrictions/restriction.hh

```diff
@@ -58,6 +59,38 @@ namespace restrictions {
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+class binary_operator;
+class conjunction;
+
+/// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
+using expression = std::variant<bool, conjunction, binary_operator>;
```

Comment:
Suggestion for a follow-up: move expressions in all their glory to cql3::expr namespace an cql3/expr subdirectory. We can alias expr::expression to expression to avoid the duplication.

Expressions will hopefully escape the restrictions cage soon and be useful everywhere.